File: netadapter/middleware/builtin/deadline.go
package builtin

// deadline.go bounds how long a call may run.

import (
	"fmt"
	"time"

	"github.com/the-protobuf-project/grpc-gateway-rs/netadapter/middleware"
)

// defaultMaxDeadline is the ceiling a client may request.
const defaultMaxDeadline = 5 * time.Minute

// Deadline bounds how long a call may run.
//
// The deadline comes from, in order: the client's Grpc-Timeout header, then the
// configured default. An adapter must always have one — an unbounded default
// turns a single slow backend into connection-pool exhaustion, and by the time
// that is visible the cause is several layers away.
//
// A client asking for longer than the maximum is capped rather than refused,
// since the request is otherwise perfectly valid.
type Deadline struct {
	// fallback is the deadline used when the client asks for none.
	fallback time.Duration

	// max is the ceiling a client may request.
	max time.Duration

	// domain is the API's error domain.
	domain string
}

// NewDeadline returns a deadline with the given default and a five-minute
// ceiling.
func NewDeadline(fallback time.Duration, domain string) *Deadline {
	return &Deadline{fallback: fallback, max: defaultMaxDeadline, domain: domain}
}

// WithMax sets the ceiling a client may request.
func (d *Deadline) WithMax(max time.Duration) *Deadline {
	d.max = max
	return d
}

// Resolve returns the deadline for one request.
func (d *Deadline) Resolve(cx *middleware.RouteCx) time.Duration {
	requested, ok := middleware.ParseGrpcTimeout(cx.Request.Header.Get("Grpc-Timeout"))
	if !ok {
		return d.fallback
	}
	return min(requested, d.max)
}

// Name implements [middleware.Interceptor].
func (*Deadline) Name() string { return "deadline" }

// OnRoute forwards the resolved deadline to the service as grpc-timeout, so the
// backend stops working on a call the adapter has already abandoned.
func (d *Deadline) OnRoute(cx *middleware.RouteCx) error {
	deadline := d.Resolve(cx)
	cx.Metadata.Append("grpc-timeout", fmt.Sprintf("%dm", deadline.Milliseconds()))
	return nil
}

// OnRequest fails the call if the deadline passed while it was in flight.
func (d *Deadline) OnRequest(cx *middleware.CallCx) error {
	if cx.Expired() {
		return cx.DeadlineError(d.domain)
	}
	return nil
}

File: transcode-rs/transcode/src/middleware/builtin/deadline.rs
//! Deadlines.

use crate::error::Result;
use crate::middleware::metadata::parse_grpc_timeout;
use crate::middleware::{CallCx, Interceptor, RouteCx};
use std::time::Duration;

/// Bounds how long a call may run.
///
/// The deadline comes from, in order: the client's `Grpc-Timeout` header, a
/// per-method override, then the configured default. A transcoder must always
/// have one — an unbounded default turns a single slow backend into
/// connection-pool exhaustion, and by the time that is visible the cause is
/// several layers away.
///
/// A client asking for longer than [`Deadline::max`] is capped rather than
/// refused, since the request is otherwise perfectly valid.
#[derive(Debug, Clone)]
pub struct Deadline {
    default: Duration,
    max: Duration,
    /// The API's error domain, stamped into the `ErrorInfo` of an expiry.
    ///
    /// Carried rather than looked up because an `ErrorInfo` naming the wrong
    /// domain is worse than useless: a caller routes on it, so a literal like
    /// `"gateway"` sends them looking for a service that does not exist.
    domain: &'static str,
}

impl Deadline {
    /// A deadline with the given default and a ceiling of five minutes.
    #[must_use]
    pub const fn new(default: Duration, domain: &'static str) -> Self {
        Self {
            default,
            max: Duration::from_secs(300),
            domain,
        }
    }

    /// Sets the ceiling a client may request.
    #[must_use]
    pub const fn with_max(mut self, max: Duration) -> Self {
        self.max = max;
        self
    }

    /// Resolves the deadline for one request.
    #[must_use]
    pub fn resolve(&self, cx: &RouteCx<'_>) -> Duration {
        cx.headers
            .get("grpc-timeout")
            .and_then(|value| value.to_str().ok())
            .and_then(parse_grpc_timeout)
            .map_or(self.default, |requested| requested.min(self.max))
    }
}

impl Interceptor for Deadline {
    fn name(&self) -> &'static str {
        "deadline"
    }

    /// Forwards the resolved deadline to the service as `grpc-timeout`, so the
    /// backend stops working on a call the transcoder has already abandoned.
    fn on_route(&self, cx: &mut RouteCx<'_>) -> Result<()> {
        let deadline = self.resolve(cx);
        cx.metadata
            .append("grpc-timeout", format!("{}m", deadline.as_millis()));
        Ok(())
    }

    /// Fails the call if the deadline passed while it was in flight.
    fn on_request(&self, cx: &mut CallCx<'_>) -> Result<()> {
        if cx.expired() {
            return Err(cx.deadline_error(self.domain));
        }
        Ok(())
    }
}
